In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import os
import datetime
import re 
from collections import Counter
print(os.listdir("../input"))

['imbd-sf']


In [2]:
data = pd.read_csv('../input/imbd-sf/data.csv')
data.head(3)

imdb_id  popularity     budget     revenue      original_title  \
0  tt0369610   32.985763  150000000  1513528810      Jurassic World   
1  tt1392190   28.419936  150000000   378436354  Mad Max: Fury Road   
2  tt2908446   13.112507  110000000   295238201           Insurgent   

                                                cast          director  \
0  Chris Pratt|Bryce Dallas Howard|Irrfan Khan|Vi...   Colin Trevorrow   
1  Tom Hardy|Charlize Theron|Hugh Keays-Byrne|Nic...     George Miller   
2  Shailene Woodley|Theo James|Kate Winslet|Ansel...  Robert Schwentke   

                      tagline  \
0           The park is open.   
1          What a Lovely Day.   
2  One Choice Can Destroy You   

                                            overview  runtime  \
0  Twenty-two years after the events of Jurassic ...      124   
1  An apocalyptic story set in the furthest reach...      120   
2  Beatrice Prior must confront her inner demons ...      119   

                                      genres  \
0  Action|Adventure|Science Fiction|Thriller   
1  Action|Adventure|Science Fiction|Thriller   
2         Adventure|Science Fiction|Thriller   

                                production_companies release_date  vote_count  \
0  Universal Studios|Amblin Entertainment|Legenda...     6/9/2015        5562   
1  Village Roadshow Pictures|Kennedy Miller Produ...    5/13/2015        6185   
2  Summit Entertainment|Mandeville Films|Red Wago...    3/18/2015        2480   

   vote_average  release_year  
0           6.5          2015  
1           7.1          2015  
2           6.3          2015

In [3]:
len(data)

1890

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1890 entries, 0 to 1889
Data columns (total 16 columns):
imdb_id                 1890 non-null object
popularity              1890 non-null float64
budget                  1890 non-null int64
revenue                 1890 non-null int64
original_title          1890 non-null object
cast                    1890 non-null object
director                1890 non-null object
tagline                 1890 non-null object
overview                1890 non-null object
runtime                 1890 non-null int64
genres                  1890 non-null object
production_companies    1890 non-null object
release_date            1890 non-null object
vote_count              1890 non-null int64
vote_average            1890 non-null float64
release_year            1890 non-null int64
dtypes: float64(2), int64(5), object(9)
memory usage: 236.4+ KB


In [5]:
data.describe()

popularity        budget       revenue      runtime   vote_count  \
count  1890.000000  1.890000e+03  1.890000e+03  1890.000000  1890.000000   
mean      1.560611  5.450696e+07  1.552890e+08   109.653439   785.666667   
std       1.849471  4.931705e+07  2.146386e+08    18.013641  1068.725880   
min       0.010335  5.000000e+06  2.033165e+06    63.000000    10.000000   
25%       0.643284  2.000000e+07  3.434007e+07    97.000000   164.000000   
50%       1.065065  3.800000e+07  8.358664e+07   106.500000   389.000000   
75%       1.804537  7.200000e+07  1.782136e+08   120.000000   937.750000   
max      32.985763  4.250000e+08  2.781506e+09   214.000000  9767.000000   

       vote_average  release_year  
count   1890.000000   1890.000000  
mean       6.140899   2007.861905  
std        0.764584      4.467929  
min        3.300000   2000.000000  
25%        5.600000   2004.000000  
50%        6.100000   2008.000000  
75%        6.600000   2012.000000  
max        8.100000   2015.000000

# Предобработка датасета

In [6]:
answer_ls = []
data['profit']=data.revenue-data.budget
def is_pos(val):
    if val>0:
        return 1
    else:
        return 0
data['is_profit']=data.profit.apply(is_pos) 

# 1. У какого фильма из списка самый большой бюджет?
Варианты ответов:
1. The Dark Knight Rises (tt1345836)
2. Spider-Man 3 (tt0413300)
3. Avengers: Age of Ultron (tt2395427)
4. The Warrior's Way	(tt1032751)
5. Pirates of the Caribbean: On Stranger Tides (tt1298650)

In [7]:
data[data.budget == data.budget.max()]

imdb_id  popularity     budget   revenue     original_title  \
491  tt1032751     0.25054  425000000  11087569  The Warrior's Way   

                                                  cast    director  \
491  Kate Bosworth|Jang Dong-gun|Geoffrey Rush|Dann...  Sngmoo Lee   

                     tagline  \
491  Assassin. Hero. Legend.   

                                              overview  runtime  \
491  An Asian assassin (Dong-gun Jang) is forced to...      100   

                                        genres      production_companies  \
491  Adventure|Fantasy|Action|Western|Thriller  Boram Entertainment Inc.   

    release_date  vote_count  vote_average  release_year     profit  is_profit  
491    12/2/2010          74           6.4          2010 -413912431          0

In [8]:
answer_ls.append(4)

# 2. Какой из фильмов самый длительный (в минутах)
1. The Lord of the Rings: The Return of the King	(tt0167260)
2. Gods and Generals	(tt0279111)
3. King Kong	(tt0360717)
4. Pearl Harbor	(tt0213149)
5. Alexander	(tt0346491)

In [9]:
data[data.runtime==data.runtime.max()]

imdb_id  popularity    budget   revenue     original_title  \
1158  tt0279111    0.469518  56000000  12923936  Gods and Generals   

                                                   cast           director  \
1158  Stephen Lang|Jeff Daniels|Robert Duvall|Kevin ...  Ronald F. Maxwell   

                                  tagline  \
1158  The nations heart was touched by...   

                                               overview  runtime  \
1158  The film centers mostly around the personal an...      214   

                 genres                production_companies release_date  \
1158  Drama|History|War  Turner Pictures|Antietam Filmworks    2/21/2003   

      vote_count  vote_average  release_year    profit  is_profit  
1158          23           5.8          2003 -43076064          0

In [10]:
answer_ls.append(2)

# 3. Какой из фильмов самый короткий (в минутах)
Варианты ответов:

1. Home on the Range	tt0299172
2. The Jungle Book 2	tt0283426
3. Winnie the Pooh	tt1449283
4. Corpse Bride	tt0121164
5. Hoodwinked!	tt0443536

In [11]:
data[data.runtime==data.runtime.min()]

imdb_id  popularity    budget   revenue   original_title  \
769  tt1449283    1.425344  30000000  14460000  Winnie the Pooh   

                                                  cast  \
769  Jim Cummings|Travis Oates|Jim Cummings|Bud Luc...   

                      director   tagline  \
769  Stephen Anderson|Don Hall  Oh Pooh.   

                                              overview  runtime  \
769  During an ordinary day in Hundred Acre Wood, W...       63   

               genres                               production_companies  \
769  Animation|Family  Walt Disney Pictures|Walt Disney Animation Stu...   

    release_date  vote_count  vote_average  release_year    profit  is_profit  
769    4/13/2011         174           6.8          2011 -15540000          0

In [12]:
answer_ls.append(3)

# 4. Какое число ближе к средней длительности фильма в датасете?

Варианты ответов:
1. 115
2. 110
3. 105
4. 120
5. 100


In [13]:
round(data.runtime.mean())

110

In [14]:
answer_ls.append(2)

# 5. Какое число ближе к медианной длительности фильма в датасете?
Варианты ответов:
1. 106
2. 112
3. 101
4. 120
5. 115

In [15]:
round(data.runtime.median())

106

In [16]:
answer_ls.append(1)

# 6. Какой самый прибыльный фильм?
Варианты ответов:
1. The Avengers	tt0848228
2. Minions	tt2293640
3. Star Wars: The Force Awakens	tt2488496
4. Furious 7	tt2820852
5. Avatar	tt0499549

In [17]:
data[data.profit==data.profit.max()]

imdb_id  popularity     budget     revenue original_title  \
239  tt0499549    9.432768  237000000  2781505847         Avatar   

                                                  cast       director  \
239  Sam Worthington|Zoe Saldana|Sigourney Weaver|S...  James Cameron   

                         tagline  \
239  Enter the World of Pandora.   

                                              overview  runtime  \
239  In the 22nd century, a paraplegic Marine is di...      162   

                                       genres  \
239  Action|Adventure|Fantasy|Science Fiction   

                                  production_companies release_date  \
239  Ingenious Film Partners|Twentieth Century Fox ...   12/10/2009   

     vote_count  vote_average  release_year      profit  is_profit  
239        8458           7.1          2009  2544505847          1

In [18]:
answer_ls.append(5)

# 7. Какой фильм самый убыточный?
Варианты ответов:
1. Supernova tt0134983
2. The Warrior's Way tt1032751
3. Flushed Away	tt0424095
4. The Adventures of Pluto Nash	tt0180052
5. The Lone Ranger	tt1210819

In [19]:
data[data.profit==data.profit.min()]

imdb_id  popularity     budget   revenue     original_title  \
491  tt1032751     0.25054  425000000  11087569  The Warrior's Way   

                                                  cast    director  \
491  Kate Bosworth|Jang Dong-gun|Geoffrey Rush|Dann...  Sngmoo Lee   

                     tagline  \
491  Assassin. Hero. Legend.   

                                              overview  runtime  \
491  An Asian assassin (Dong-gun Jang) is forced to...      100   

                                        genres      production_companies  \
491  Adventure|Fantasy|Action|Western|Thriller  Boram Entertainment Inc.   

    release_date  vote_count  vote_average  release_year     profit  is_profit  
491    12/2/2010          74           6.4          2010 -413912431          0

In [20]:
answer_ls.append(2)

# 8. У скольких фильмов из датасета объем сборов оказался выше бюджета?
Варианты ответов:
1. 1478
2. 1520
3. 1241
4. 1135
5. 1398

In [21]:
len(data[data.profit>0])

1478

In [22]:
answer_ls.append(1)

# 9. Какой фильм оказался самым кассовым в 2008 году?
Варианты ответов:
1. Madagascar: Escape 2 Africa	tt0479952
2. Iron Man	tt0371746
3. Kung Fu Panda	tt0441773
4. The Dark Knight	tt0468569
5. Mamma Mia!	tt0795421

In [23]:
data[data.profit == data[data.release_year==2008].profit.max()]

imdb_id  popularity     budget     revenue   original_title  \
600  tt0468569    8.466668  185000000  1001921825  The Dark Knight   

                                                  cast           director  \
600  Christian Bale|Michael Caine|Heath Ledger|Aaro...  Christopher Nolan   

             tagline                                           overview  \
600  Why So Serious?  Batman raises the stakes in his war on crime. ...   

     runtime                       genres  \
600      152  Drama|Action|Crime|Thriller   

                                  production_companies release_date  \
600  DC Comics|Legendary Pictures|Warner Bros.|Syncopy    7/16/2008   

     vote_count  vote_average  release_year     profit  is_profit  
600        8432           8.1          2008  816921825          1

In [24]:
answer_ls.append(4)

# 10. Самый убыточный фильм за период с 2012 по 2014 (включительно)?
Варианты ответов:
1. Winter's Tale	tt1837709
2. Stolen	tt1656186
3. Broken City	tt1235522
4. Upside Down	tt1374992
5. The Lone Ranger	tt1210819


In [25]:
int_data = data[(data.release_year>=2012) & (data.release_year<=2014)]
int_data[int_data.profit==int_data.profit.min()]

imdb_id  popularity     budget   revenue   original_title  \
1246  tt1210819     1.21451  255000000  89289910  The Lone Ranger   

                                                   cast        director  \
1246  Johnny Depp|Armie Hammer|William Fichtner|Hele...  Gore Verbinski   

                      tagline  \
1246  Never Take Off the Mask   

                                               overview  runtime  \
1246  The Texas Rangers chase down a gang of outlaws...      149   

                        genres  \
1246  Action|Adventure|Western   

                                   production_companies release_date  \
1246  Walt Disney Pictures|Jerry Bruckheimer Films|I...     7/3/2013   

      vote_count  vote_average  release_year     profit  is_profit  
1246        1607           6.0          2013 -165710090          0

In [26]:
answer_ls.append(5)

# 11. Какого жанра фильмов больше всего?
Варианты ответов:
1. Action
2. Adventure
3. Drama
4. Comedy
5. Thriller

In [27]:
def genre_stat(data):
    com=[]
    for k in range(len(data.genres)):
        cur_genre = data.iloc[k].genres.split('|')
        for j in range(len(cur_genre)):
            com.append(cur_genre[j])
    com = pd.Series(com) 
    return com.value_counts()   
print(genre_stat(data).index[0],'->',genre_stat(data).iloc[0])

Drama -> 782


In [28]:
answer_ls.append(3)

# 12. Какого жанра среди прибыльных фильмов больше всего?
Варианты ответов:
1. Drama
2. Comedy
3. Action
4. Thriller
5. Adventure

In [29]:
data_prof=data[data.profit>0]
print(genre_stat(data_prof).index[0],'->',genre_stat(data_prof).iloc[0])

Drama -> 560


In [30]:
answer_ls.append(1)

# 13. Кто из режиссеров снял больше всего фильмов?
Варианты ответов:
1. Steven Spielberg
2. Ridley Scott 
3. Steven Soderbergh
4. Christopher Nolan
5. Clint Eastwood

In [31]:
info = data.director.value_counts()
print(info.index[0],'->',info.iloc[0])

Steven Soderbergh -> 13


In [32]:
answer_ls.append(3)

# 14. Кто из режиссеров снял больше всего Прибыльных фильмов?
1. Steven Spielberg
2. Ridley Scott 
3. Steven Soderbergh
4. Christopher Nolan
5. Clint Eastwood

In [33]:
data_prof=data[data.profit>0]
info = data_prof.director.value_counts()
print(info.index[0],'->',info.iloc[0])

Ridley Scott -> 12


In [34]:
answer_ls.append(2)

# 15. Кто из режиссеров принес больше всего прибыли?
Варианты ответов:
1. Steven Spielberg
2. James Cameron
3. Christopher Nolan
4. David Yates
5. Peter Jackson


In [35]:
info = data.groupby(["director"])["profit"].sum().sort_values(ascending=False)
print(info.index[0],'->',info.iloc[0])

Peter Jackson -> 5202593685


In [36]:
answer_ls.append(5)

# 16. Какой актер принес больше всего прибыли?
Варианты ответов:
1. Emma Watson
2. Johnny Depp
3. Michelle Rodriguez
4. Orlando Bloom
5. Rupert Grint

In [37]:
def prof_info(data):
    c_list=[]
    prof_list=[]
    for k in range(len(data.cast)):
        cur_cast = data.iloc[k].cast.split('|')
        cur_prof = data.iloc[k].profit
        for j in range(len(cur_cast)):
            c_list.append(cur_cast[j])
            prof_list.append(cur_prof)
    new_data = {'Actor':c_list,'Profit':prof_list} 
    df = pd.DataFrame(new_data) 
    return df

prof_data = prof_info(data)
info=prof_data.groupby(["Actor"])["Profit"].sum().sort_values(ascending=False)
print(info.index[0],'->',info.iloc[0])

Emma Watson -> 6666245597


In [38]:
answer_ls.append(1)

# 17. Какой актер принес меньше всего прибыли в 2012 году?
Варианты ответов:
1. Nicolas Cage
2. Danny Huston
3. Kirsten Dunst
4. Jim Sturgess
5. Sami Gayle

In [39]:
data_2012=data[(data.release_year==2012)]
prof_data = prof_info(data_2012)
info = prof_data.groupby(["Actor"])["Profit"].sum().sort_values(ascending=True)
print(info.index[0],'->',info.iloc[0])

Kirsten Dunst -> -68109207


In [40]:
answer_ls.append(3)

# 18. Какой актер снялся в большем количестве высокобюджетных фильмов? Примечание: в фильмах, где бюджет выше среднего по данной выборке.
1. Tom Cruise
2. Mark Wahlberg 
3. Matt Damon
4. Angelina Jolie
5. Adam Sandler

In [41]:
def actor_info(data):
    c_list=[]
    for k in range(len(data.cast)):
        cur_cast = data.iloc[k].cast.split('|')
        for j in range(len(cur_cast)):
            c_list.append(cur_cast[j])
    ds = pd.Series(c_list) 
    return ds

data_high=data[data.budget>data.budget.mean()]
info = actor_info(data_high).value_counts()
print(info.index[0],'->',info.iloc[0])

Matt Damon -> 18


In [42]:
answer_ls.append(3)

# 19. В фильмах какого жанра больше всего снимался Nicolas Cage?  
Варианты ответа:
1. Drama
2. Action
3. Thriller
4. Adventure
5. Crime

In [43]:
def is_NC(data):
    res=[]    
    for k in range(len(data.cast)):
        cur_cast = data.iloc[k].cast.split('|')
        if 'Nicolas Cage' in cur_cast:
            res.append(1)
        else:
            res.append(0)
    return pd.Series(res) 
 
data['is_NC']=is_NC(data)
data_NC = data[data.is_NC==1]
info = genre_stat(data_NC)
print(info.index[0],'->',info.iloc[0])

Action -> 17


In [44]:
answer_ls.append(2)

# 20. Какая студия сняла больше всего фильмов?
Варианты ответа:
1. Universal Pictures (Universal)
2. Paramount Pictures
3. Columbia Pictures
4. Warner Bros
5. Twentieth Century Fox Film Corporation

In [45]:
def prod_comp_stat(data):
    com=[]
    for k in range(len(data.production_companies)):
        cur_comp = data.iloc[k].production_companies.split('|')
        for j in range(len(cur_comp)):
            com.append(cur_comp[j])
    return pd.Series(com).value_counts() 

info = prod_comp_stat(data)
print(info.index[0],'->',info.iloc[0])  

Universal Pictures -> 173


In [46]:
answer_ls.append(1)

# 21. Какая студия сняла больше всего фильмов в 2015 году?
Варианты ответа:
1. Universal Pictures
2. Paramount Pictures
3. Columbia Pictures
4. Warner Bros
5. Twentieth Century Fox Film Corporation

In [47]:
data_2015=data[data.release_year==2015]
info = prod_comp_stat(data_2015)
print(info.index[0],'->',info.iloc[0])  

Warner Bros. -> 12


In [48]:
answer_ls.append(4)

# 22. Какая студия заработала больше всего денег в жанре комедий за все время?
Варианты ответа:
1. Warner Bros
2. Universal Pictures (Universal)
3. Columbia Pictures
4. Paramount Pictures
5. Walt Disney

In [49]:
def is_comedy(data):
    res=[]    
    for k in range(len(data.genres)):
        cur_genre = data.iloc[k].genres.split('|')
        if 'Comedy' in cur_genre:
            res.append(1)
        else:
            res.append(0)
    return pd.Series(res) 

col_comedy = is_comedy(data)  
data['is_comedy']=col_comedy

def studio_info(data):
    s_list=[]
    prof_list=[]
    for k in range(len(data.production_companies)):
        cur_s = data.iloc[k].production_companies.split('|')
        cur_prof = data.iloc[k].profit
        for j in range(len(cur_s)):
            s_list.append(cur_s[j])
            prof_list.append(cur_prof)
    new_data = {'Company':s_list,'Profit':prof_list} 
    df = pd.DataFrame(new_data) 
    return df

data_comedy = data[data.is_comedy==1]
comp_data = studio_info(data_comedy)

info=comp_data.groupby(["Company"])["Profit"].sum().sort_values(ascending=False)
print(info.index[0],'->',info.iloc[0])  

Universal Pictures -> 8961545581


In [50]:
answer_ls.append(2)

# 23. Какая студия заработала больше всего денег в 2012 году?
Варианты ответа:
1. Universal Pictures (Universal)
2. Warner Bros
3. Columbia Pictures
4. Paramount Pictures
5. Lucasfilm

In [51]:
data_2012=data[data.release_year==2012]
comp_data = studio_info(data_2012)
info=comp_data.groupby(["Company"])["Profit"].sum().sort_values(ascending=False)
print(info.index[0],'->',info.iloc[0])  

Columbia Pictures -> 2501406608


In [52]:
answer_ls.append(3)

# 24. Самый убыточный фильм от Paramount Pictures
Варианты ответа:

1. K-19: The Widowmaker tt0267626
2. Next tt0435705
3. Twisted tt0315297
4. The Love Guru tt0811138
5. The Fighter tt0964517

In [53]:
def is_PP(data):
    res=[]    
    for k in range(len(data.production_companies)):
        cur_comp = data.iloc[k].production_companies.split('|')
        if 'Paramount Pictures' in cur_comp:
            res.append(1)
        else:
            res.append(0)
    return pd.Series(res) 
 
data['is_par_pic']=is_PP(data) 
data_PP = data[data.is_par_pic==1]
sort_data = data_PP.sort_values(by=['profit'])
print(sort_data.iloc[0].original_title,'->',sort_data.iloc[0].profit)

K-19: The Widowmaker -> -64831034


In [54]:
answer_ls.append(1)

# 25. Какой Самый прибыльный год (заработали больше всего)?
Варианты ответа:
1. 2014
2. 2008
3. 2012
4. 2002
5. 2015

In [55]:
info = data.groupby(["release_year"])["profit"].sum().sort_values(ascending=False)
print(info.index[0],'->',info.iloc[0])  

2015 -> 18668572378


In [56]:
answer_ls.append(5)

# 26. Какой Самый прибыльный год для студии Warner Bros?
Варианты ответа:
1. 2014
2. 2008
3. 2012
4. 2010
5. 2015

In [57]:
answers = { 
            "1": 2014,
            "2": 2008,
            "3": 2012,
            "4": 2010,
            "5": 2015
          }
def is_WB(data):
    res=[]    
    for k in range(len(data.production_companies)):
        cur_comp = data.iloc[k].production_companies.split('|')
        if ('Warner Bros.' in cur_comp) | ('Warner Bros. Pictures' in cur_comp): 
            res.append(1)
        else:
            res.append(0)
    return pd.Series(res) 

data['is_warn_broth']=is_WB(data)  
data_WB = data[data.is_warn_broth==1]
info = data_WB.groupby(["release_year"])["profit"].sum().sort_values(ascending=False)
years = list(info.index)

for k in range(len(years)):
        for key, value in answers.items():
            if value==years[k]:
                print(years[k])         

2008
2014
2010
2012
2015


In [58]:
answer_ls.append(2)

# 27. В каком месяце за все годы суммарно вышло больше всего фильмов?
Варианты ответа:
1. Январь
2. Июнь
3. Декабрь
4. Сентябрь
5. Май

In [59]:
def month_col(data):
    m_list=[]
    for k in range(len(data.release_date)):
        cur=data.iloc[k].release_date.split('/')
        m_list.append(datetime.date(int(cur[2]),int(cur[0]),int(cur[1])).strftime('%b'))
    return pd.Series(m_list)
data['month']=month_col(data)
info = data.month.value_counts()
print(info.index[0],'->',info.iloc[0]) 

Sep -> 227


In [60]:
answer_ls.append(4)

# 28. Сколько суммарно вышло фильмов летом? (за июнь, июль, август)
Варианты ответа:
1. 345
2. 450
3. 478
4. 523
5. 381

In [61]:
data_summer=data[data.month.isin(['Jun','Jul','Aug'])]
len(data_summer)

450

In [62]:
answer_ls.append(2)

# 29. Какой режисер выпускает (суммарно по годам) больше всего фильмов зимой?
Варианты ответов:
1. Steven Soderbergh
2. Christopher Nolan
3. Clint Eastwood
4. Ridley Scott
5. Peter Jackson

In [63]:
data_winter=data[data.month.isin(['Jan','Dec','Feb'])]
info = data_winter.director.value_counts()
print(info.index[0],'->',info.iloc[0]) 

Peter Jackson -> 7


In [64]:
answer_ls.append(5)

# 30. Какой месяц чаще всего самый прибыльный в году?
Варианты ответа:
1. Январь
2. Июнь
3. Декабрь
4. Сентябрь
5. Май

In [65]:
# list of years
years = data.release_year.value_counts()
years = list(years.index)

best_mon=[]
for k in range(len(years)):
    data_year = data[data.release_year == years[k]]
    info = data_year.groupby(["month"])["profit"].sum().sort_values(ascending=False)
    best_mon.append(info.index[0])
info = pd.Series(best_mon).value_counts() 
print(info.index[0],'->',info.iloc[0])  

Jun -> 7


In [66]:
answer_ls.append(2)

# 31. Названия фильмов какой студии в среднем самые длинные по количеству символов?
Варианты ответа:
1. Universal Pictures (Universal)
2. Warner Bros
3. Jim Henson Company, The
4. Paramount Pictures
5. Four By Two Productions

In [67]:
def film_info(data):
    s_list=[]
    len_list=[]
    for k in range(len(data.production_companies)):
        cur_s = data.iloc[k].production_companies.split('|')
        cur_len = len(data.iloc[k].original_title)
        for j in range(len(cur_s)):
            s_list.append(cur_s[j])
            len_list.append(cur_len)
    new_data = {'Company':s_list,'Length':len_list} 
    df = pd.DataFrame(new_data) 
    return df

film_len = film_info(data)
info = film_len.groupby(["Company"])["Length"].mean().sort_values(ascending=False)
print(info.index[0],'->',info.iloc[0]) 

Four By Two Productions -> 83.0


In [68]:
answer_ls.append(5)

# 32. Названия фильмов какой студии в среднем самые длинные по количеству слов?
Варианты ответа:
1. Universal Pictures (Universal)
2. Warner Bros
3. Jim Henson Company, The
4. Paramount Pictures
5. Four By Two Productions

In [69]:
def count_words(phrase):
    phrase = re.sub(r'(?=[MDCLXVI])M*D?C{0,4}L?X{0,4}V?I{0,4}$', '', phrase)
    phrase = re.sub(r'[^a-z ]+', '', phrase.lower())
    return len(phrase.split())

def film_words(data):
    s_list=[]
    len_list=[]
    for k in range(len(data.production_companies)):
        cur_s = data.iloc[k].production_companies.split('|')
        cur_len = count_words(data.iloc[k].original_title)
        for j in range(len(cur_s)):
            s_list.append(cur_s[j])
            len_list.append(cur_len)
    new_data = {'Company':s_list,'Amount':len_list} 
    df = pd.DataFrame(new_data) 
    return df

film_name_words = film_words(data)
info = film_name_words.groupby(["Company"])["Amount"].mean().sort_values(ascending=False)
print(info.index[0],'->',info.iloc[0]) 

Four By Two Productions -> 12.0


In [70]:
answer_ls.append(5)

# 33. Сколько разных слов используется в названиях фильмов?(без учета регистра)
Варианты ответа:
1. 6540
2. 1002
3. 2461
4. 28304
5. 3432

In [71]:
def n_words(phrase):    
    phrase = re.sub(r'(?=[MDCLXVI])M*D?C{0,4}L?X{0,4}V?I{0,4}$', '', phrase)
    phrase = re.sub(r'[^a-z ]+', '', phrase.lower())
    return phrase.split()

def all_words(data):
    w_list=[]
    for k in range(len(data)):
        cur_words = n_words(data.iloc[k].original_title)
        for j in range(len(cur_words)):
            w_list.append(cur_words[j])
    return pd.Series(w_list)

words = all_words(data)
words.nunique()

2306

In [72]:
answer_ls.append(3) 

# 34. Какие фильмы входят в 1 процент лучших по рейтингу?
Варианты ответа:
1. Inside Out, Gone Girl, 12 Years a Slave
2. BloodRayne, The Adventures of Rocky & Bullwinkle
3. The Lord of the Rings: The Return of the King, Upside Down
4. 300, Lucky Number Slevin
5. Upside Down, 300, Inside Out, The Lord of the Rings: The Return of the King

In [73]:
sort_data = data.sort_values(by=['vote_average'],ascending=False)
n = round(len(sort_data)/100)
titles=sort_data.iloc[0:n].original_title
answers={
  "1": ['Inside Out', 'Gone Girl', '12 Years a Slave'],
  "2": ['BloodRayne', 'The Adventures of Rocky & Bullwinkle'],
  "3": ['The Lord of the Rings: The Return of the King','Upside Down'],
  "4": ['300', 'Lucky Number Slevin'],
  "5": ['Upside Down', '300', 'Inside Out', 'The Lord of the Rings: The Return of the King']
}

for key, value in answers.items():
        if set(value).issubset(list(titles)):
            print(key,':',value)

1 : ['Inside Out', 'Gone Girl', '12 Years a Slave']


In [74]:
answer_ls.append(1)

# 35. Какие актеры чаще всего снимаются в одном фильме вместе
Варианты ответа:
1. Johnny Depp & Helena Bonham Carter
2. Hugh Jackman & Ian McKellen
3. Vin Diesel & Paul Walker
4. Adam Sandler & Kevin James
5. Daniel Radcliffe & Rupert Grint

In [75]:
import itertools
answers = { 
            "1": ['Johnny Depp', 'Helena Bonham Carter'],
            "2": ['Hugh Jackman', 'Ian McKellen'],
            "3": ['Vin Diesel', 'Paul Walker'],
            "4": ['Adam Sandler', 'Kevin James'],
            "5": ['Daniel Radcliffe', 'Rupert Grint']
          }

actors = list(itertools.chain(*list(answers.values())))

def get_AF_dict(data,actors):
    t_list=[]
    for k in range(len(data)):
        cur_act = data.iloc[k].cast.split('|')
        cur_id = data.iloc[k].imdb_id
        for j in range(len(cur_act)):
            if cur_act[j] in actors:
                t_list.append([cur_act[j],cur_id])
    res = pd.DataFrame(t_list, columns =['Actor', 'Film_id']) 
    return dict(res.groupby('Actor')['Film_id'].apply(list))

def max_pair(data):  
    indx=[]
    max_val=0;
    keys=list(data.keys())
    for k in range(len(keys)-1):
        a = data[keys[k]]
        for j in range(k+1,len(keys)):
            b = data[keys[j]]
            n = len(set(a).intersection(b))
            if n>max_val:
                max_val = n 
                indx=[k,j]
    return [keys[indx[0]], keys[indx[1]], max_val]

res = max_pair(get_AF_dict(data,actors))
ans = res[:2] 

for key,val in answers.items():
    if val == ans:
        print(key,':',ans)

5 : ['Daniel Radcliffe', 'Rupert Grint']


In [76]:
answer_ls.append(5)

# 36. У какого из режиссеров самый высокий процент фильмов со сборами выше бюджета?
Варианты ответа:
1. Quentin Tarantino
2. Steven Soderbergh
3. Robert Rodriguez
4. Christopher Nolan
5. Clint Eastwood

In [77]:
answers = { 
            "1": ['Quentin Tarantino'],
            "2": ['Steven Soderbergh'],
            "3": ['Robert Rodriguez'],
            "4": ['Christopher Nolan'],
            "5": ['Clint Eastwood']
          }

def prob(data):
    return 100*sum(data)/len(data)
     
directors = list(itertools.chain(*list(answers.values())))

data_dir = data[data.director.isin(directors)] 

val = data_dir.groupby('director')['is_profit'].apply(prob)

val.sort_values(ascending=False)

director
Quentin Tarantino    100.000000
Christopher Nolan    100.000000
Robert Rodriguez      85.714286
Clint Eastwood        83.333333
Steven Soderbergh     76.923077
Name: is_profit, dtype: float64

In [78]:
#additional investigation
def actors(data):
    a_list=[]
    for k in range(len(data)):
        cur_act = data.iloc[k].cast.split('|')
        for j in range(len(cur_act)):
            a_list.append(cur_act[j])
    return pd.Series(a_list)

def actors_prof(data,actors):
    a_list=[]
    for k in range(len(data)):
        cur_act=data.iloc[k].cast.split('|')
        cur_prof=data.iloc[k].is_profit
        for j in range(len(cur_act)):
            if cur_act[j] in actors:
                a_list.append([cur_act[j],cur_prof])
    res = pd.DataFrame(a_list, columns =['Actor', 'Film_prof']) 
    return dict(res.groupby('Actor')['Film_prof'].apply(list))

def get_profit(data):
    for key,val in data.items():
        data[key]=sum(val)/len(val)
    return data

def get_weig_profit(actors,profit):
        names=list(actors.keys())
        res=0;
        for name in names:
            res=res+profit[name]*actors[name]
        return res/sum(list(actors.values()))

# collect data for each director
data_CN = data[data.director=='Christopher Nolan'] 
data_QT = data[data.director=='Quentin Tarantino']

# create dict actor:frequency 
actors_CN = dict(actors(data_CN).value_counts())
actors_QT  = dict(actors(data_QT).value_counts())

# create list with all actors 
actors_com=[]
actors_com.append(list(actors_CN.keys()))
actors_com.append(list(actors_QT.keys()))
actors_com = list(itertools.chain(*actors_com))

# create dict actor:film_prof for all films
profit = actors_prof(data,actors_com)

# individual profit for each actor
profit_ind = get_profit(profit)

# compute weight-average probability of profit for actors 
QT_prob = get_weig_profit(actors_QT,profit)
CN_prob = get_weig_profit(actors_CN,profit)
print('Tarantino:',QT_prob,'; Nolan:',CN_prob)

Tarantino: 0.8861878306878307 ; Nolan: 0.8177341061716064


In [79]:
answer_ls.append(1)

# Submission

In [80]:
len(answer_ls)

36

In [81]:
pd.DataFrame({'Id':range(1,len(answer_ls)+1), 'Answer':answer_ls}, columns=['Id', 'Answer'])

Id  Answer
0    1       4
1    2       2
2    3       3
3    4       2
4    5       1
5    6       5
6    7       2
7    8       1
8    9       4
9   10       5
10  11       3
11  12       1
12  13       3
13  14       2
14  15       5
15  16       1
16  17       3
17  18       3
18  19       2
19  20       1
20  21       4
21  22       2
22  23       3
23  24       1
24  25       5
25  26       2
26  27       4
27  28       2
28  29       5
29  30       2
30  31       5
31  32       5
32  33       3
33  34       1
34  35       5
35  36       1

## ATTENTION
Шаблон ноутбука не полностью соответствовал тесту на платформе. Кроме несовпадающих постановок вопросов имело место несовпадение порядка предлагаемых ответов. Я привел шаблон в соотвествие с платформой. Пожалуйста, обратите внимание на это при проверке. 

По результатам теста у меня два неправильных ответа (вопросы 26, 36), с чем я не совсем согласен. Хотелось бы получить подробное объяснение. Кроме того, меня очень интересует вид регулярного выражения для поиска слов и их количества в названиях фильмов (вопросы 32, 33). 